README.md

# motorcortex-LeonSans

## [Demo](https://kissmybutton.github.io/motorcortex-leonsans/demo/)


## Installation

```bash
$ npm install --save @kissmybutton/motorcortex-leonsans
# OR
$ yarn add @kissmybutton/motorcortex-leonsans
```

## Loading

```javascript
import {loadPlugin} from "@kissmybutton/motorcortex/";
import LeonPlugin from "@kissmybutton/motorcortex-leonsans";
const LeonSans = loadPlugin(LeonPlugin);
```

# LeonSans Clip
By the use of the plugin's `Clip` you can instantiate / render a text anywhere in your DOM. The Incident will render a canvas which will host your text. 

The `Clip` takes a number of attributes (listed below) that define both the canvas and the rendered text characteristics and it can accept an unlimited number of `LeonSans` Incidents that can animate the rendered text.

## Example

```javascript
const clipName = new LeonSans.Clip(
  {
    canvasId: "leonObj",
    sw: 800,
    sh: 600,
    text: "abcdefghijklmn\nopqrstuvwxyz",
    weight: 100,
    round: true,
    multiply: true,
    drawing: "colorful",
    color: [["#64d3ce", "#2a92ce82", "#2a92ce40", "#2a92ce2e", "#2a92ce00"]],
    colorful: ["#64d3ce", "#2a92ce82", "#2a92ce40", "#2a92ce2e", "#2a92ce00"],
    pathGap: -1,
    patternWidth: 1,
    patternHeight: 10,
    speed: 1,
    tracking: 20,
    leading: 10,
    size: 80
  },
  {
    selector: ".test",
   
  }
);
```

## Supported Attrs explained

| Name        | Are           | Values  |
| ------------- |:-------------:| -----:|
| canvasId  | the id of the rendered canvas that contains the text. This id will be used by all `LeonIncident` Incidents, added to the `Clip` timeline, via their `selector` property | string | 
| sw      | width of clip canvas (how many pixels) | all positive numbers |
| sh      | height of clip canvas (how many pixels) | all positive numbers |
| text |  the text to display |  string  |
| weight |  The weight of the font | number 1 - 900   |
| round |  round the corners of font if they are not path | boolean |
| multiply |  applying multiply effect when the font has more then 1 colors |  boolean  |
| drawing |  type of the font effect |  one of: `drawing`, `pattern`, `colorful`, `colorPattern`  |
| color |  The colors of the characters  |  array of arrays of colors   |
| colorful |  The list of colors of each character. Only applicable if drawing property is `colorfull`   |  array of colors   |
| pathGap |  The gap between each coordinate of the points on a line of each character |  number |
| patternWidth: |  the width of the lines when drawing property is `pattern`  |  number    |
| patternHeight |  the height of the lines when drawing property is `pattern`   |  number  |
| speed |  speed of color changing  when drawing property is `colorPattern`  |  number, from 0 to 10   |
| tracking |  The spacing between the characters of a block of text   |  number    |
| leading |  The distance between each line of text  |  number    |
| size |  size of text   |  number |





# LeonIncident

## Example

```javascript
const nameOfIncident = new LeonSans.LeonIncident(
  {
    animatedAttrs: {
      LeonAtrs: {
        completion_rate: 1,
        weight: 600,
        // tracking: 20,
        leading: 10,
        // size: 90,
        // pathGap: 1,
        patternWidth: 100
        // patternHeight: 90
      }
    },
    initialValues: {
      LeonAtrs: {
        completion_rate: 0,
        weight: 100,
        // tracking: 10,
        // leading: 1,
        size: 80,
        // pathGap: 0,
        patternWidth: 1
        // patternHeight: 10
      }
    }
  },
  {
    duration: 2000,
    selector: "!#leonObj"
  }
);
```

## Supported Attrs explained
The `LeonIncident` attrs takes on the `animatedAttrs` object just one composite attribute: `LeonAtrs`. The supported attributes of this composite attribute are:

| Name        | Are           | Values  |
| ------------- |:-------------:| -----:|
| completion_rate |  The rate of text completion |  num from 0 to 1  |
| weight |  The weight of the font: 1 - 900 |  number  |
| pathGap |  The gap between each coordinate of the points on a line of each character |  number |
| patternWidth: |  the width of the lines when drawing property is `pattern`  |  number    |
| patternHeight |  the height of the lines when drawing property is `pattern`   |  number  |
| tracking |  The spacing between the characters of a block of text   |  number    |
| leading |  The distance between each line of text  |  number    |
| size |  size of text   |  number |

Along with the attributes, all `LeonIncident` incidents must take on their props the `selector` key which targets the canvas id of their parent Clip via the convention: `!#<canvasId>`

# License
[MIT License](https://opensource.org/licenses/MIT)


  
  
[![Kiss My Button](https://presskit.kissmybutton.gr/logos/kissmybutton-logo-small.png)](https://kissmybutton.gr)









CHANGELOG.md
# Changelog

### 1.0.13 (2021-04-16)


### Features

* change package json for interoperable OS support ([d11b4c6](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/d11b4c626bc9ecbfdb6c7f67d1e862f0de5fc69b))
* first implementation of the boilerpalte ([2c4bfe8](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/2c4bfe85e42ca7917792dc8dfba9ad67c8cd5bc6))


### Bug Fixes

* add clip validation rules ([f0a3d11](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/f0a3d11ed3987ac85c0450a536add87ba7cc6369))
* add validation rules ([355e0c2](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/355e0c21653fc8036e3c82c274931f2f1bbc5122))
* **build demo:** build demo ([b2525ac](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/b2525ac82bdf3acddf6ed88d7e340c413e78d875))
* drawing bug fix ([c5b5bee](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/c5b5beed4f00b3dc81427f59a18c753ded9cf3f8))
* fix demo ([9caade0](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/9caade05f64e91356c6920180a3db4401b848225))
* fix demo ([8928ac1](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/8928ac13dead6ee9bd2c956eebc3e2b67cba39a9))
* rollup implementation ([69d6451](https://www.github.com/kissmybutton/motorcortex-leonsans/commit/69d6451c8138824b43f172d07e88d457dec60802))


src/LeonSans.js
import MC from "@kissmybutton/motorcortex";
import leonSans from "./src/index.js";
let cValue = 0;
const PI2 = 2 * Math.PI;

class LeonSans extends MC.BrowserClip {
  get html() {
    return ` <canvas> </canvas>`;
  }

  onAfterRender() {
    this.cValue = 0;
    const canvas = this.context.getElements("canvas")[0];
    this.sw = this.attrs.sw;
    this.sh = this.attrs.sh;
    const pixelRatio = 2;

    this.ctx = canvas.getContext("2d");

    canvas.width = this.sw * pixelRatio;
    canvas.height = this.sh * pixelRatio;
    canvas.style.width = this.sw + "px";
    canvas.style.height = this.sh + "px";
    this.ctx.scale(pixelRatio, pixelRatio);
    const leon = new leonSans({
      text: this.attrs.text,
      color: this.attrs.color,
      weight: this.attrs.weight,
      isPath: true,
      pathGap:
        this.attrs.pathGap || this.attrs.drawing === "colorPattern" ? -1 : 0,
      isWave: this.attrs.drawing === "wave" ? true : false,
      tracking: this.attrs.tracking,
      leading: this.attrs.leading,
      size: this.attrs.size || 80,
      patternWidth: 10,
      patternHeight: 90,
      colorful: this.attrs.colorful
    });

    // leon.grid(this.ctx);

    this.attrs.round
      ? (this.ctx.lineCap = "round")
      : (this.ctx.lineCap = "butt");

    this.attrs.multiply
      ? (this.ctx.globalCompositeOperation = "multiply")
      : (this.ctx.globalCompositeOperation = "source-over");

    this.setCustomEntity(this.attrs.canvasId, {
      leon: leon,
      ctx: this.ctx,
      sw: this.sw,
      sh: this.sh,
      drawing: this.attrs.drawing,
      pathGap: this.attrs.pathGap,
      patternWidth: this.attrs.patternWidth,
      patternHeight: this.attrs.patternHeight
    });

    // switch (this.attrs.drawing) {
    //   case "colorful":
    //     leon.drawColorful(this.ctx);
    //     break;
    //   case "pattern":
    //     leon.pattern(
    //       this.ctx,
    //       this.attrs.patternWidth,
    //       this.attrs.patternHeight
    //     );
    //     break;
    //   case "wave":
    //     break;
    //   default:
    //     leon.draw(this.ctx);
    // }

    this.leon = leon;
  }
  onAfterProgress() {
    if (this.attrs.drawing === "colorPattern") {
      this.ctx.clearRect(0, 0, this.sw, this.sh);
      this.ctx.lineWidth = 0.2;
      const w = this.attrs.patternWidth;
      const total = this.leon.data.length;
      let i,
        pos,
        no = 0;
      let d, j, j_total;
      for (i = 0; i < total; i++) {
        d = this.leon.data[i].paths;
        j_total = Math.round(d.length * this.leon.drawing[i].value);
        for (j = 0; j < j_total; j++) {
          pos = d[j];
          this.ctx.fillStyle = this.randomColor(no);
          this.ctx.strokeStyle = this.randomColor(no);
          this.ctx.beginPath();
          this.ctx.arc(pos.x, pos.y, w, 0, PI2);
          this.ctx.stroke();
          no += 1;
        }
      }

      cValue -= this.attrs.speed;
    }
  }
  randomColor(no) {
    return "hsl(" + (no + cValue) + "," + "70%," + "50%)";
  }
}

export default LeonSans;
